xdsl/dialects/riscv.py

```diff
@@ -74,47 +78,63 @@ def __init__(self, flags: None | Sequence[FastMathFlag] | Literal["none", "fast"
         super().__init__(flags)
 
 
-class RISCVRegisterType(Data[str], TypeAttribute, ABC):
+class RISCVRegisterType(ParametrizedAttribute, TypeAttribute, ABC):
     """
     A RISC-V register type.
     """
 
-    _unallocated: ClassVar[Self | None] = None
+    index: ParameterDef[IntAttr | NoneAttr]
+    spelling: ParameterDef[StringAttr]
+
+    def __init__(self, spelling: str):
+        super().__init__(self._parameters_from_spelling(spelling))
+
+    @classmethod
+    def _parameters_from_spelling(
+        cls, spelling: str
+    ) -> tuple[IntAttr | NoneAttr, StringAttr]:
+        """
+        Returns the parameter list required to construct a register instance from the given spelling.
+        """
+        index_attr = NoneAttr()
+        index = cls.abi_index_by_name().get(spelling)
+        if index is not None:
+            index_attr = IntAttr(index)
+        return index_attr, StringAttr(spelling)
 
     @classmethod
     def unallocated(cls) -> Self:
-        if cls._unallocated is None:
-            cls._unallocated = cls("")
-        return cls._unallocated
+        return cls("")
```

Comment:
While we're out it, could you please make this an abstract method, and in the subclasses return either `Registers.UNALLOCATED_INT` or `Registers.UNALLOCATED_FLOAT`?